geomesa-fs/geomesa-fs-datastore/src/main/scala/org/locationtech/geomesa/fs/FileSystemFeatureStore.scala

```diff
@@ -89,14 +88,27 @@ class FileSystemFeatureStore(val storage: FileSystemStorage,
     }
   }
 
-  override def getBoundsInternal(query: Query): ReferencedEnvelope = storage.getMetadata.getEnvelope
   override def buildFeatureType(): SimpleFeatureType = sft
+
+  override def getBoundsInternal(query: Query): ReferencedEnvelope = {
+    import org.locationtech.geomesa.utils.geotools.{CRS_EPSG_4326, wholeWorldEnvelope}
+    val partitions = storage.getPartitions(query.getFilter).iterator
+    if (!partitions.hasNext) { wholeWorldEnvelope } else {
+      val envelope = new ReferencedEnvelope(CRS_EPSG_4326)
+      while (partitions.hasNext) {
+        envelope.expandToInclude(partitions.next.bounds())
+      }
+      envelope
+    }
+  }
+
   override def getCountInternal(query: Query): Int = {
-    if (query == Query.ALL || query.getFilter.isInstanceOf[IncludeFilter]) {
-      storage.getMetadata.getFeatureCount.toInt
-    } else {
-      -1L
-    }.toInt
+    val partitions = storage.getPartitions(query.getFilter).iterator
```

Comment:
Are there any sane options for memoizing this?

Reply:
i don't think that this should be a particularly expensive call

Reply:
the partitions are already cached in memory, so it's just iterating through them to match the filter